File: src/main/java/com/jesperancinha/xmladder/instruction/XmlAdderInstruction.java
package com.jesperancinha.xmladder.instruction;

import java.util.Map;
import java.util.Set;

/**
 * Created by joaofilipesabinoesperancinha on 16-02-16.
 */
public class XmlAdderInstruction {

    private final Map<String, String> attributesToAdd;

    private final String attributeValueToReplace;

    private final Map<String, String> attributeKeyValueToReplaceOrAdd;

    private final Set<String> attributeKeysToDelete;

    private final Map<String, Map<String, String>> siblingsToAdd;

    private final Map<String, String> siblingToAddValues;


    /**
     * Only attributesToAdd attribute will be available on version Zero
     *
     * @param attributeValueToReplace
     * @param attributesToAdd
     * @param attributeKeyValueToReplaceOrAdd
     * @param siblingsToAdd
     * @param siblingToAddValues
     */
    public XmlAdderInstruction( //
                                String attributeValueToReplace, //
                                Map<String, String> attributesToAdd, //
                                Map<String, String> attributeKeyValueToReplaceOrAdd, //
                                Set<String> attributeKeyToDelete, //
                                Map<String, Map<String, String>> siblingsToAdd, //
                                Map<String, String> siblingToAddValues //
    ) {
        this.attributesToAdd = attributesToAdd;
        this.attributeValueToReplace = attributeValueToReplace;
        this.attributeKeyValueToReplaceOrAdd = attributeKeyValueToReplaceOrAdd;
        this.siblingsToAdd = siblingsToAdd;
        this.siblingToAddValues = siblingToAddValues;
        this.attributeKeysToDelete = attributeKeyToDelete;
    }

    public void addAddAttribute(String name, String value) {
        attributesToAdd.put(name, value);
    }

    public void addDeleteAttribute(String name) {
        attributeKeysToDelete.add(name);
    }

    public Map<String, String> getAttributesToAdd() {
        return attributesToAdd;
    }

    public Set<String> getAttributeKeysToDelete() {
        return attributeKeysToDelete;
    }
}

File: app/src/main/kotlin/org/jesperancinha/xml/adder/instruction/XmlAdderInstruction.kt
package org.jesperancinha.xml.adder.instruction

/**
 * Created by joaofilipesabinoesperancinha on 16-02-16.
 */
class XmlAdderInstruction
/**
 * Only attributesToAdd attribute will be available on version Zero
 *
 * @param attributeValueToReplace
 * @param attributesToAdd
 * @param attributeKeyValueToReplaceOrAdd
 * @param siblingsToAdd
 * @param siblingToAddValues
 */(
    private val attributeValueToReplace: String? = null,
    internal val attributesToAdd: MutableMap<String, String>,
    private val attributeKeyValueToReplaceOrAdd: Map<String, String>? = null,
    internal val attributeKeysToDelete: MutableSet<String>,
    private val siblingsToAdd: Map<String, Map<String, String>>? = null,
    private val siblingToAddValues: Map<String, String>? = null
) {
    fun addAddAttribute(name: String, value: String?) = value?.let { attributesToAdd[name] = value }

    fun addDeleteAttribute(name: String?) = name?.let {
        attributeKeysToDelete.add(name)
    }

    fun getAttributesToAdd(): Map<String, String> {
        return attributesToAdd
    }

    fun getAttributeKeysToDelete(): Set<String> {
        return attributeKeysToDelete
    }
}